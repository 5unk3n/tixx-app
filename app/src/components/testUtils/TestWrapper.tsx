import { NavigationContainer } from '@react-navigation/native'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import React, { PropsWithChildren } from 'react'
import { PaperProvider } from 'react-native-paper'

import { theme } from '@/theme/defaultTheme'

const queryClient = new QueryClient({
	defaultOptions: {
		queries: {
			retry: false,
			staleTime: Infinity
		}
	}
})

export const TestWrapper = ({ children }: PropsWithChildren) => {
	return (
		<PaperProvider theme={theme.dark}>
			<QueryClientProvider client={queryClient}>
				<NavigationContainer>{children}</NavigationContainer>
			</QueryClientProvider>
		</PaperProvider>
	)
}
