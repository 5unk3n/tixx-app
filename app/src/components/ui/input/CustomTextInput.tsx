import React, { useRef } from 'react'
import {
	Pressable,
	TextInput,
	StyleSheet,
	View,
	TextInputProps,
	StyleProp,
	ViewStyle
} from 'react-native'
import { HelperText } from 'react-native-paper'

import { useCustomTheme } from '@/hooks/useCustomTheme'

import { CustomText } from '../display/CustomText'

export interface CustomTextInputProps extends TextInputProps {
	className?: string
	label?: string
	style?: StyleProp<ViewStyle>
	containerStyle?: StyleProp<ViewStyle>
	errorMessage?: string
	disabled?: boolean
	right?: JSX.Element
}

export default function CustomTextInput({
	style,
	label,
	errorMessage,
	right,
	disabled,
	numberOfLines,
	containerStyle,
	...props
}: CustomTextInputProps) {
	const ref = useRef<TextInput>(null)
	const { colors, fonts } = useCustomTheme()

	return (
		<View style={style}>
			{label && (
				<Pressable className="pl-2 pb-2" onPress={() => ref.current?.focus()}>
					<CustomText variant="caption1Regular" className="text-grayscale-500">
						{label}
					</CustomText>
				</Pressable>
			)}
			<View
				style={[
					styles.container,
					{ backgroundColor: colors.surfaceVariant },
					numberOfLines
						? styles.containerFullHeight
						: styles.containerDefaultHeight,
					containerStyle
				]}
			>
				<TextInput
					ref={ref}
					className={`flex-1 px-5 font-medium text-base leading-md tracking-tight ${disabled ? 'text-grayscale-400' : 'text-onPrimary'} ${numberOfLines ? 'py-4' : 'py-2'}`}
					placeholderTextColor={colors.grayscale[600]}
					editable={!disabled}
					numberOfLines={numberOfLines}
					{...props}
				/>
				{right && (
					<View className="grow-0 items-center justify-center py-2 px-2">
						{right}
					</View>
				)}
			</View>
			<HelperText
				type="error"
				visible={!!errorMessage}
				style={[styles.text, fonts.caption1Regular, { color: colors.error }]}
			>
				{errorMessage}
			</HelperText>
		</View>
	)
}

const styles = StyleSheet.create({
	text: {
		textAlign: 'right',
		position: 'absolute',
		paddingHorizontal: 8,
		paddingVertical: 0,
		right: 0,
		bottom: -27
	},
	container: {
		flexDirection: 'row',
		borderRadius: 8
	},
	containerDefaultHeight: {
		height: 48
	},
	containerFullHeight: {
		height: '100%'
	}
})
