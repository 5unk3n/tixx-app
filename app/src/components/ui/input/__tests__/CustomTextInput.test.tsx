import { describe, it, expect, jest } from '@jest/globals'
import { render, screen, fireEvent } from '@testing-library/react-native'
import React from 'react'
import { TextInput } from 'react-native'

import { TestWrapper } from '../../../testUtils/TestWrapper'
import CustomTextInput, { CustomTextInputProps } from '../CustomTextInput'

describe('CustomTextInput', () => {
	it('should handle text input correctly', () => {
		const onChangeTextMock = jest.fn()
		render(
			<CustomTextInput
				value="initial"
				onChangeText={onChangeTextMock}
				accessibilityLabel="text-input"
			/>,
			{ wrapper: TestWrapper }
		)

		const input = screen.getByLabelText('text-input')
		fireEvent.changeText(input, 'new text')
		expect(onChangeTextMock).toHaveBeenCalledWith('new text')
	})

	it('should focus the input when the label is pressed', () => {
		// Mock the focus method
		const mockFocus = jest.fn()
		jest.spyOn(TextInput.prototype, 'focus').mockImplementation(mockFocus)

		render(
			<CustomTextInput label="이메일" value="" onChangeText={() => {}} />,
			{ wrapper: TestWrapper }
		)

		fireEvent.press(screen.getByText('이메일'))
		expect(mockFocus).toHaveBeenCalledTimes(1)

		// Restore original implementation
		mockFocus.mockRestore()
	})

	describe('should render correctly based on props', () => {
		const testCases: Array<Partial<CustomTextInputProps>> = [
			{ errorMessage: '오류 메시지' },
			{ right: <TextInput testID="right-element" /> },
			{ disabled: true },
			{ numberOfLines: 3, multiline: true },
			{ label: '레이블', right: <></>, errorMessage: '오류' }
		]

		testCases.forEach((props, i) => {
			it(`renders correctly for case ${i + 1}`, () => {
				const { toJSON } = render(
					<CustomTextInput value="test" onChangeText={() => {}} {...props} />,
					{ wrapper: TestWrapper }
				)
				expect(toJSON()).toMatchSnapshot()
			})
		})
	})
})
