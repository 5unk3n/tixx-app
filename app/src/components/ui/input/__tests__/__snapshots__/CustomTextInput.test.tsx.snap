// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTextInput should render correctly based on props renders correctly for case 1 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          [
            {
              "borderRadius": 8,
              "flexDirection": "row",
            },
            {
              "backgroundColor": "#1B1B1B",
            },
            {
              "height": 48,
            },
            undefined,
          ]
        }
      >
        <TextInput
          editable={true}
          onChangeText={[Function]}
          placeholderTextColor="#585858"
          style={
            [
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              },
              {
                "fontFamily": "Pretendard-Medium",
                "fontWeight": "500",
              },
              {
                "fontSize": 16,
              },
              {
                "letterSpacing": -0.32,
              },
              {
                "color": "#FFFFFF",
              },
              {
                "paddingBottom": 8,
                "paddingTop": 8,
              },
            ]
          }
          value="test"
        />
      </View>
      <Text
        collapsable={false}
        maxFontSizeMultiplier={1.5}
        onLayout={[Function]}
        style={
          {
            "bottom": -27,
            "color": "#F2F862",
            "fontFamily": "Pretendard-Regular",
            "fontSize": 12,
            "fontWeight": "400",
            "letterSpacing": -0.02,
            "lineHeight": 14,
            "opacity": 1,
            "paddingHorizontal": 8,
            "paddingVertical": 0,
            "position": "absolute",
            "right": 0,
            "textAlign": "right",
            "transform": [
              {
                "translateY": -0,
              },
            ],
            "writingDirection": "ltr",
          }
        }
      >
        오류 메시지
      </Text>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomTextInput should render correctly based on props renders correctly for case 2 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          [
            {
              "borderRadius": 8,
              "flexDirection": "row",
            },
            {
              "backgroundColor": "#1B1B1B",
            },
            {
              "height": 48,
            },
            undefined,
          ]
        }
      >
        <TextInput
          editable={true}
          onChangeText={[Function]}
          placeholderTextColor="#585858"
          style={
            [
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              },
              {
                "fontFamily": "Pretendard-Medium",
                "fontWeight": "500",
              },
              {
                "fontSize": 16,
              },
              {
                "letterSpacing": -0.32,
              },
              {
                "color": "#FFFFFF",
              },
              {
                "paddingBottom": 8,
                "paddingTop": 8,
              },
            ]
          }
          value="test"
        />
        <View
          style={
            [
              {
                "flexGrow": 0,
              },
              {
                "alignItems": "center",
              },
              {
                "justifyContent": "center",
              },
              {
                "paddingBottom": 8,
                "paddingTop": 8,
              },
              {
                "paddingLeft": 8,
                "paddingRight": 8,
              },
            ]
          }
        >
          <TextInput
            testID="right-element"
          />
        </View>
      </View>
      <Text
        collapsable={false}
        maxFontSizeMultiplier={1.5}
        onLayout={[Function]}
        style={
          {
            "bottom": -27,
            "color": "#F2F862",
            "fontFamily": "Pretendard-Regular",
            "fontSize": 12,
            "fontWeight": "400",
            "letterSpacing": -0.02,
            "lineHeight": 14,
            "opacity": 0,
            "paddingHorizontal": 8,
            "paddingVertical": 0,
            "position": "absolute",
            "right": 0,
            "textAlign": "right",
            "transform": [],
            "writingDirection": "ltr",
          }
        }
      />
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomTextInput should render correctly based on props renders correctly for case 3 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          [
            {
              "borderRadius": 8,
              "flexDirection": "row",
            },
            {
              "backgroundColor": "#1B1B1B",
            },
            {
              "height": 48,
            },
            undefined,
          ]
        }
      >
        <TextInput
          editable={false}
          onChangeText={[Function]}
          placeholderTextColor="#585858"
          style={
            [
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              },
              {
                "fontFamily": "Pretendard-Medium",
                "fontWeight": "500",
              },
              {
                "fontSize": 16,
              },
              {
                "letterSpacing": -0.32,
              },
              {
                "color": "#A0A2A2",
              },
              {
                "paddingBottom": 8,
                "paddingTop": 8,
              },
            ]
          }
          value="test"
        />
      </View>
      <Text
        collapsable={false}
        maxFontSizeMultiplier={1.5}
        onLayout={[Function]}
        style={
          {
            "bottom": -27,
            "color": "#F2F862",
            "fontFamily": "Pretendard-Regular",
            "fontSize": 12,
            "fontWeight": "400",
            "letterSpacing": -0.02,
            "lineHeight": 14,
            "opacity": 0,
            "paddingHorizontal": 8,
            "paddingVertical": 0,
            "position": "absolute",
            "right": 0,
            "textAlign": "right",
            "transform": [],
            "writingDirection": "ltr",
          }
        }
      />
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomTextInput should render correctly based on props renders correctly for case 4 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          [
            {
              "borderRadius": 8,
              "flexDirection": "row",
            },
            {
              "backgroundColor": "#1B1B1B",
            },
            {
              "height": "100%",
            },
            undefined,
          ]
        }
      >
        <TextInput
          editable={true}
          multiline={true}
          numberOfLines={3}
          onChangeText={[Function]}
          placeholderTextColor="#585858"
          style={
            [
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              },
              {
                "fontFamily": "Pretendard-Medium",
                "fontWeight": "500",
              },
              {
                "fontSize": 16,
              },
              {
                "letterSpacing": -0.32,
              },
              {
                "color": "#FFFFFF",
              },
              {
                "paddingBottom": 16,
                "paddingTop": 16,
              },
            ]
          }
          value="test"
        />
      </View>
      <Text
        collapsable={false}
        maxFontSizeMultiplier={1.5}
        onLayout={[Function]}
        style={
          {
            "bottom": -27,
            "color": "#F2F862",
            "fontFamily": "Pretendard-Regular",
            "fontSize": 12,
            "fontWeight": "400",
            "letterSpacing": -0.02,
            "lineHeight": 14,
            "opacity": 0,
            "paddingHorizontal": 8,
            "paddingVertical": 0,
            "position": "absolute",
            "right": 0,
            "textAlign": "right",
            "transform": [],
            "writingDirection": "ltr",
          }
        }
      />
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomTextInput should render correctly based on props renders correctly for case 5 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "paddingLeft": 8,
            },
            {
              "paddingBottom": 8,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "#FFFFFF",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "Pretendard-Regular",
                  "fontSize": 12,
                  "fontWeight": "400",
                  "letterSpacing": -0.02,
                  "lineHeight": 14,
                },
                [
                  {
                    "color": "#7E7D7D",
                  },
                ],
              ],
            ]
          }
        >
          레이블
        </Text>
      </View>
      <View
        style={
          [
            {
              "borderRadius": 8,
              "flexDirection": "row",
            },
            {
              "backgroundColor": "#1B1B1B",
            },
            {
              "height": 48,
            },
            undefined,
          ]
        }
      >
        <TextInput
          editable={true}
          onChangeText={[Function]}
          placeholderTextColor="#585858"
          style={
            [
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              },
              {
                "fontFamily": "Pretendard-Medium",
                "fontWeight": "500",
              },
              {
                "fontSize": 16,
              },
              {
                "letterSpacing": -0.32,
              },
              {
                "color": "#FFFFFF",
              },
              {
                "paddingBottom": 8,
                "paddingTop": 8,
              },
            ]
          }
          value="test"
        />
        <View
          style={
            [
              {
                "flexGrow": 0,
              },
              {
                "alignItems": "center",
              },
              {
                "justifyContent": "center",
              },
              {
                "paddingBottom": 8,
                "paddingTop": 8,
              },
              {
                "paddingLeft": 8,
                "paddingRight": 8,
              },
            ]
          }
        />
      </View>
      <Text
        collapsable={false}
        maxFontSizeMultiplier={1.5}
        onLayout={[Function]}
        style={
          {
            "bottom": -27,
            "color": "#F2F862",
            "fontFamily": "Pretendard-Regular",
            "fontSize": 12,
            "fontWeight": "400",
            "letterSpacing": -0.02,
            "lineHeight": 14,
            "opacity": 1,
            "paddingHorizontal": 8,
            "paddingVertical": 0,
            "position": "absolute",
            "right": 0,
            "textAlign": "right",
            "transform": [
              {
                "translateY": -0,
              },
            ],
            "writingDirection": "ltr",
          }
        }
      >
        오류
      </Text>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
